aip/general/0129.md

```diff
@@ -0,0 +1,71 @@
+
+## Guidance
+
+### Server Side Defaults
+There are instances where a service has to ensure that a value always exists even if the client chooses not to specify one. For example: a client may create a caching service instance without specifying the software version to be used for the service. Such a scenario requires the service to pick a version. The service designer has a number of ways to handle server side defaults.
+
+#### Dynamic Defaults
+When a default value of a field may change at any point in time such a field is considered to have a dynamic default. In general, dynamic fields can be: 
+* generated (UUID)
+* allocated (dynamic IP address)
+* assigned (most recent software package version)
+
+An attribute with a dynamic default ***must*** be expressed as two fields in the API:
+* a mutable field that can be optionally set by the user
+* an immutable `OUTPUT_ONLY` field that records the final value decided on by the API service
+
+Example:
+```proto
+message VirtualMachine {
+  …
+  string ip_address = 4;
+  string provisioned_ip_address = 5 [
+    (google.api.field_behavior) = OUTPUT_ONLY
+  ];
+}
+```
+
+#### Static Defaults
+Service ***may*** choose to designate a value to be an equivalent to an empty value. This type of default is convenient for values that don't change very often since a change to a static default is considered a non-backward-compatible change within the same major API version as specified in [AIP-180](aip-180). To designate a value as a static default a service must add google.api.default annotation to the field definition. Annotating a field with google.api.default specifies machine-readable documentation of the field's behavior, allowing clients to understand default behavior and the meaning of an empty field.
+
+Example:
+```proto
+message VM {
+ string sku = 1 [(google.api.default) = "foo"];
+}
+```
+
+### Mutable fields
+API Services ***must not*** modify mutable fields.
+
+### Semantic Strings Handling
+For a number of string transmitted value types an API service may be required to convert strings into other formats. For example, a UUID value may be transmitted from the client as an uppercased string `"3BBAB074-4E18-41F1-AD21-B7358B3DB0CE"`, stored in a database as a sequence of bytes and transmitted back to the client as a lowercase string `"3bbab074-4e18-41f1-ad21-b7358b3db0ce"`. The semantic meaning of such a string doesn't change however state-driven clients have no way of determining such differences. To avoid such mismatching values:
+
+* service must define and validate a single representation of a semantic string. 
+* service must return an error when a client submitted value doesn't match defined format.
```

Comment:
This feels like it hurts imperative clients for the sake of declarative clients. I can absolutely see why declarative clients would want this, but I'd prefer not to put a burden on other clients.

I don't remember seeing a discussion on this requirement before - which doesn't mean it hasn't happened, of course. If you could point me (internally) to any docs around this, that would be helpful.

Reply:
This isn't just about declarative clients. When the service mutates values, that makes it difficult to determine equivalence and correctness without detailed, hardcoded knowledge of the service, which may even need to be reverse-engineered through trial and error.

It affects any clients that use the state in a general way to make decisions or to compare with a "golden" representation. This includes declarative tools, but also policy engines (e.g., Open Policy Agent), security response agents, controllers, data-driven test frameworks, recommendation tools, LLMs, and so on. 

It even affects "imperative" clients that aren't manually coded. Let's say a GUI has a form to update a particular value. The user enters "x", which is allowed, and the service changes it to "X". Did the update succeed? It depends on how the GUI determines completion/success and whether a success response was recorded. 

Input/output values are not a well supported concept in many tools outside the Kubernetes ecosystem.

Reply:
The easiest thing to do in many cases is to just make validation more strict -- disallow values that would be transformed.

Reply:
That's easiest for *some* clients, but definitely not for all.

If I were writing a GUI performing an update, I'd expect that an update RPC which succeeds has performed the requested update. If I want to find the canonical representation, I can look at the response. I don't need to include that canonicalization code in my application.

With the restriction here, *every* application has to perform canonicalization unless their "natural" input happens to be the canonical form. I strongly expect the correctness of that canonicalization code to be a hit-and-miss affair. Even the fact that canonicalization is required is likely not to be obvious to any but the most diligent app developers until they first receive an error.

There's a *potential* mitigation for client library users if we have a fixed set of canonicalized string formats: we can provide canonicalization code once per language. (We could even *potentially* apply canonicalization transparently in the client, but that feels a bit too much like "magic" for me.)

It's possible that strict validation is the least-worst option, but I think it merits further discussion - and I'd like to acknowledge the trade-offs within the rationale.

Reply:
I think +1 for some set of defined data types. We could start there, I'm sure that catches 90% of the situations where value normalization occurs.

The other 10% would be bespoke normalization, and validation is a good fallback. 

I don't know how opinionated we have to be here really. handling normalization or defaulting of user-specified values can be broken down into several options, of which the service teams can choose one:

1. choose a well-known data type, enabling some normalization or validation client-side. Don't normalize on the server, validated and reject.
1. don't normalize values anywhere, instead validate and reject malformed ones.
1. normalize or default values, and provide a paired OUTPUT_ONLY value to see the final result.
1. choose a specific default value, document it, and never change it.

Reply:
Clients can be categorized on multiple axes: special case vs generic, manually constructed vs automated, static vs dynamic, management vs data plane, operation-driven vs state-driven, connected vs at rest, and so on.

The case you're describing sounds like a special-case, manually constructed, static, connected, operation-driven management client. To understand the canonicalization for that specific field of that resource type of that service, a human would have had to read documentation or proto comments, analyze gcloud source code, and/or discover the behavior through trial and error. 

Clients that are passing data values from one system to another generally would have to transform and reformat the data. That's normal.

Clients that are taking input values from users could just request values within the valid range. That's also normal.

Services transforming resource attributes provided by the client violates the principle of least surprise as well as obstructing generic, automated, state-driven clients. The behavior is different from typical fields, which are not transformed by services, and currently is mechanically undiscoverable.

We could standardize a "strict" parameter on Create and Update to require pre-canonicalized values, but I'm unconvinced that that complexity is necessary.

Another approach could be to handle them the same way as dynamic defaults, with an "effective_" output-only field.

Reply:
Yeah, this would be a case where there's a "Value Type" is of a more specific class than the in-transit/in-memory field type holding it. Especially because of cases where strength of typing in-transit on different protocols already varies - ability to GET/PUT json representation means map/array/string/number is always the most extreme (and universally supported) case. 

I like the approach of thinking in terms of normalization and validation first.

**Validation** is really an invariant. The client doesn't need to validate, but the service will of course only accept a valid resource reference for a ref string field. I think the robustness principle would apply there "be conservative in what you do, be liberal in what you accept from others" --- so accepting upper-case/lower-case UUID, etc, would be better than requiring the client to mirror exact marshaling of fields in all cases.

It is true you could say there's one way to solve "value-entropy vs representation-entropy" by saying there's one single acceptable representation for each valid value... But that would put callers in a position where they could be fighting the format of their underlying inputs... e.g. to_lower(guidgen())... So I'd suggest letting the representation entropy be larger and accept more than one well-formed representation of the same value.

**Normalization** is where it gets interesting. With (un)marshalling going from representation-space to value-space and back to representation-space ---- the normalization that happens may not even be intentional. The only way to avoid any normalization is to accept the highest fidelity representation all the way from client to backend, and validate incoming but only normalize server-side at a point past what will ever ever round-trip. e.g. store strings and only convery to uuid/timestamp in what's passed to control plane.

Which is a lot... And also re "...be conservative in what you do..." you might expect that the service's job _is_ to have an extremely consistent set of representations... e.g. if you are writing client code to do evaluate policy over resources - it would be a kindness to have a field in the same representation every time rather than preserving the differences in representation that are not differences in value...

**Equivalence** is the last bit - where normalization causes problems is in value-equality, aka representation-equivalence... For a client which must compare two values for equivalence, what operator must it use?

The advantage of the service opting to heavy normalization in responses is it help --- when you have two responses --- to use naive string equality on fields to compare them...

In the case of a convergent client that's reconciling or doing drift detection --- that's really the case where it needs to know how do do value-space equivalence. Which means, rather than string==string, "what equality operator should this field be compared using?"

TL;DR suggestion...

* services should accept the broader set of representations of a field value in requests
* services should produce the same representation for the a field value in responses
* annotation must indicate when a string field contains a more specific kinds of value
* converging/annealing clients must implement comparison operators for kind/default annotations
* aip needs to define about a dozen a-priori kinds

(IMO dynamic kinds, regex value extraction, all out of scope. complexity explodes, diminishing returns. etc.)

See also... [7.3](https://datatracker.ietf.org/doc/html/draft-bhutton-json-schema-validation-00#section-7.3).  Defined Formats https://datatracker.ietf.org/doc/html/draft-bhutton-json-schema-validation-00#section-7.3
 For a starting point of "kind" annotations... You'd also consider a "resource_reference" to also be an example of where we've already added a "kind" marker onto a string field.

Reply:
Added a set of defined data types for strings